In [2]:
from pydataset import data

## When the instructions say to load a dataset, you can pass the name of the dataset as a string to the data function to load the dataset. You can also view the documentation for the data set by passing the show_doc keyword argument.

data('mpg', show_doc=True) -- view the documentation for the dataset --

mpg = data('mpg') -- load the dataset and store it in a variable --

In [6]:
data('mpg', show_doc=True) 

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




In [4]:
mpg = data('mpg')

In [5]:
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

## All the datasets loaded from the pydataset library will be pandas dataframes.
## Copy the code from the lesson to create a dataframe full of student grades.

In [8]:
import pandas as pd
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

grades_df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})

type(grades_df)

pandas.core.frame.DataFrame

## Create a column named passing_english that indicates whether each student has a passing grade in english.

In [11]:
grades_df['passing_english'] = grades_df['english'] >= 70

In [14]:
grades_df

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
11     Alan    92       62       72            False

Sort the english grades by the passing_english column. How are duplicates handled?
Duplicates handled by index sorting

In [16]:
grades_df.sort_values('passing_english')

name  math  english  reading  passing_english
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
11     Alan    92       62       72            False
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True

## Sort the english grades first by passing_english and then by student name. All the students that are failing english should be first, and within the students that are failing english they should be ordered alphabetically. The same should be true for the students passing english.

In [21]:
grades_df.sort_values(['passing_english', 'name']) 
# does sort first by passing_english, then further sorts by name.

name  math  english  reading  passing_english
11     Alan    92       62       72            False
8    Albert    92       62       87            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True
1      Jane    88       79       67             True
5      John    79       76       93             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
2     Suzie    94       74       95             True

## Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.

In [22]:
grades_df.sort_values(['passing_english', 'english'])

name  math  english  reading  passing_english
8    Albert    92       62       87            False
11     Alan    92       62       72            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
2     Suzie    94       74       95             True
5      John    79       76       93             True
1      Jane    88       79       67             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True

## Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.

In [48]:
grades_df['overall_grade'] = round((grades_df.math + grades_df.english + grades_df.reading) / 3, 2)
grades_df

name  math  english  reading  passing_english  overall_grade
0     Sally    62       85       80             True          75.67
1      Jane    88       79       67             True          78.00
2     Suzie    94       74       95             True          87.67
3     Billy    98       96       88             True          94.00
4       Ada    77       92       98             True          89.00
5      John    79       76       93             True          82.67
6    Thomas    82       64       81            False          75.67
7     Marie    93       63       90            False          82.00
8    Albert    92       62       87            False          80.33
9   Richard    69       80       94             True          81.00
10    Isaac    92       99       93             True          94.67
11     Alan    92       62       72            False          75.33

# 2. Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:

In [6]:
data('mpg', show_doc=True) 

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




In [4]:
mpg = data('mpg')

In [51]:
mpg.head(3)

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact

## How many rows and columns are there? 234 rows, 11 columns

In [56]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


In [5]:
mpg.shape

(234, 11)

## What are the data types of each column?

In [59]:
mpg.dtypes

manufacturer     object
model            object
displ           float64
year              int64
cyl               int64
trans            object
drv              object
cty               int64
hwy               int64
fl               object
class            object
dtype: object

## Summarize the dataframe with .info and .describe

In [62]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


In [63]:
mpg.describe()

displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

## Rename the cty column to city.

In [71]:
mpg.rename(columns={'cty': 'city'}, inplace=True)

## Rename the hwy column to highway.

In [72]:
mpg.rename(columns={'hwy': 'highway'}, inplace=True)

## Do any cars have better city mileage than highway mileage? No

In [78]:
mpg[mpg['city'] > mpg['highway']] # --> returns empty df

Empty DataFrame
Columns: [city, highway]
Index: []

## Create a column named mileage_difference this column should contain the difference between highway and city mileage for each car.

In [83]:
mpg['mileage_difference'] = mpg.highway - mpg.city

## Which car (or cars) has the highest mileage difference? Honda Civic, Volkswagon New Beetle

In [91]:
mpg.nlargest(1, 'mileage_difference', keep='all')

manufacturer       model  displ  year  cyl     trans drv  city  highway  \
107        honda       civic    1.8  2008    4  auto(l5)   f    24       36   
223   volkswagen  new beetle    1.9  1999    4  auto(l4)   f    29       41   

    fl       class  mileage_difference  
107  c  subcompact                  12  
223  d  subcompact                  12

## Which compact class car has the lowest highway mileage? The best?

In [98]:
mpg[mpg['class'] == 'compact'].nsmallest(1, 'highway', keep='all') # --> Lowest

manufacturer  model  displ  year  cyl     trans drv  city  highway fl  \
220   volkswagen  jetta    2.8  1999    6  auto(l4)   f    16       23  r   

       class  mileage_difference  
220  compact                   7

In [99]:
mpg[mpg['class'] == 'compact'].nlargest(1, 'highway', keep='all') # --> Best

manufacturer  model  displ  year  cyl       trans drv  city  highway fl  \
213   volkswagen  jetta    1.9  1999    4  manual(m5)   f    33       44  d   

       class  mileage_difference  
213  compact                  11

## Create a column named average_mileage that is the mean of the city and highway mileage.

In [103]:
mpg['average_mileage'] = (mpg.city + mpg.highway) / 2

## Which dodge car has the best average mileage? The worst?

In [109]:
mpg[mpg['manufacturer'] == 'dodge'].nlargest(1, 'average_mileage', keep='all')

manufacturer        model  displ  year  cyl     trans drv  city  highway  \
38        dodge  caravan 2wd    2.4  1999    4  auto(l3)   f    18       24   

   fl    class  mileage_difference  average_mileage  
38  r  minivan                   6             21.0

In [110]:
mpg[mpg['manufacturer'] == 'dodge'].nsmallest(1, 'average_mileage', keep='all')

manufacturer                model  displ  year  cyl       trans drv  city  \
55        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   
60        dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
66        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
70        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   

    highway fl   class  mileage_difference  average_mileage  
55       12  e  pickup                   3             10.5  
60       12  e     suv                   3             10.5  
66       12  e  pickup                   3             10.5  
70       12  e  pickup                   3             10.5

## Load the Mammals dataset. Read the documentation for it, and use the data to answer these questions:

In [122]:
data('Mammals', show_doc=True)

Mammals

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Garland(1983) Data on Running Speed of Mammals

### Description

Observations on the maximal running speed of mammal species and their body
mass.

### Usage

    data(Mammals)

### Format

A data frame with 107 observations on the following 4 variables.

weight

Body mass in Kg for "typical adult sizes"

speed

Maximal running speed (fastest sprint velocity on record)

hoppers

logical variable indicating animals that ambulate by hopping, e.g. kangaroos

specials

logical variable indicating special animals with "lifestyles in which speed
does not figure as an important factor": Hippopotamus, raccoon (Procyon),
badger (Meles), coati (Nasua), skunk (Mephitis), man (Homo), porcupine
(Erithizon), oppossum (didelphis), and sloth (Bradypus)

### Details

Used by Chappell (1989) and Koenker, Ng and Portnoy (1994) to illustrate the
fitting of piecewise linear curves.

### Source

Garland, T. (

In [7]:
mammals = data('Mammals')

In [124]:
mammals.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 107 entries, 1 to 107
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   weight    107 non-null    float64
 1   speed     107 non-null    float64
 2   hoppers   107 non-null    bool   
 3   specials  107 non-null    bool   
dtypes: bool(2), float64(2)
memory usage: 2.7 KB


## How many rows and columns are there? 107 rows, 4 columns

## What are the data types? float64, bool

In [125]:
mammals.dtypes

weight      float64
speed       float64
hoppers        bool
specials       bool
dtype: object

## Summarize the dataframe with .info and .describe

In [126]:
mammals.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 107 entries, 1 to 107
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   weight    107 non-null    float64
 1   speed     107 non-null    float64
 2   hoppers   107 non-null    bool   
 3   specials  107 non-null    bool   
dtypes: bool(2), float64(2)
memory usage: 2.7 KB


In [127]:
mammals.describe()

weight       speed
count   107.000000  107.000000
mean    278.688178   46.208411
std     839.608269   26.716778
min       0.016000    1.600000
25%       1.700000   22.500000
50%      34.000000   48.000000
75%     142.500000   65.000000
max    6000.000000  110.000000

## What is the the weight of the fastest animal? 55

In [134]:
mammals.nlargest(1, 'speed', keep='all')[['weight']]

weight
53    55.0

## What is the overal percentage of specials? 9.35%

In [143]:
mammals.specials.value_counts(True) * 100

False    90.654206
True      9.345794
Name: specials, dtype: float64

## How many animals are hoppers that are above the median speed? 7
## What percentage is this? 6.54%

In [167]:
mammals[(mammals.hoppers == True) & (mammals.speed > mammals.speed.median())]

weight  speed  hoppers  specials
96      4.6   64.0     True     False
97      4.4   72.0     True     False
98      4.0   72.0     True     False
99      3.5   56.0     True     False
100     2.0   64.0     True     False
101     1.9   56.0     True     False
102     1.5   50.0     True     False

In [13]:
fast_hoppers = len(mammals[(mammals.hoppers == True) & (mammals.speed > mammals.speed.median())]) / len(mammals)

In [14]:
fast_hoppers = round((fast_hoppers * 100), 2)
fast_hoppers

6.54